zip-0221.rst

```diff
@@ -0,0 +1,585 @@
+
+Given the performance of Blake2b, we expect this validation cost to be negligible. However, it seems prudent to benchmark potential MMR implementations during the implementation process. Should the validation cost be higher than expected, there are several potential mitigations, e.g. holding recently seen nodes in memory after a reorg.
+
+Generally, header commitments have no impact on privacy. However, FlyClient has additional security and privacy implications. Because FlyClient is a motivating factor for this ZIP, it seems prudent to include a brief overview. A more in-depth security analysis of FlyClient should be performed before designing a FlyClient-based light client ecosystem for ZCash.
+
+FlyClient, like all light clients, requires a connection to a light client server. That server may collect information about client requests, and may use that information to attempt to deanonymize clients. However, because FlyClient proofs are non-interactive and publicly verifiable, they could be shared among many light clients after the initial server interaction.
+
+FlyClient proofs are probabilistic. When properly constructed, there is negligible probability that a dishonest chain commitment will be accepted by the verifier. The security analysis assumes adversary mining power is bounded by a known fraction of combined mining power of honest nodes, and cannot drop or tamper with messages between client and full nodes. It also assumes the client is connected to at least one full node and knows the genesis block. However, these security properties have not been examined closely in chain models with rapidly-adjusting difficulty.
+
+
+
+Specification
+--------------
+.. image:: https://i.imgur.com/hhRyI99.png
+    :alt: zcash_MMR
+
+*Fig 1. MMR commitment scheme*
+
+The leaves of the MMR at block ``B_n`` are hash commitments to the header data and metadata of each previous block ``B_x, ..., B_(n-1)``, where ``x`` is the block height of the most recent network upgrade. We extend the standard MMR to allow metadata to propagate upwards through the tree by either summing the metadata of both children, or inheriting the metadata of a specific child as necessary. This allows us to create efficient proofs of selected properties of a range of blocks without transmitting the entire range of blocks or headers.
+
+Tree Node specification
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Unless otherwise noted, all hashes use BLAKE2b-256 with the personalization field set to ``'ZcashHistory' || CONSENSUS_BRANCH_ID``. ``CONSENSUS_BRANCH_ID`` is the little-endian encoding of ``BRANCH_ID`` for the epoch of the block containing the commitment. [#zip-0200]_ Which is to say, each node in the tree commits to the consensus branch that produced it.
```

Comment:
`CONSENSUS_BRANCH_ID` serialised as uint32 to produce actual personalization? 
So personalization will be exactly 16 bytes ("ZcashHistory".length + 4)

or as compact?
and then personalization will be 13..21 bytes ("ZcashHistory".length + encoding of `CONSENSUS_BRANCH_ID`)

I guess this should be specified here!

Reply:
serialized as uint32 so the personalization field is exactly 16 bytes.

See here for another example:
https://github.com/zcash/zips/blob/master/zip-0243.rst